lrlex/src/lib/lexer.rs

```diff
@@ -310,30 +310,18 @@ impl<'lexer, 'input: 'lexer, StorageT: Copy + Eq + Hash + PrimInt + Unsigned>
             );
         }
 
-        fn surrounding_line_off<StorageT>(
-            lexer: &LRNonStreamingLexer<StorageT>,
-            i: usize,
-        ) -> (usize, usize) {
-            if i > lexer.s.len() {
-                panic!("Offset {} exceeds known input length {}", i, lexer.s.len());
-            }
-
-            if lexer.newlines.is_empty() {
-                return (0, lexer.s.len());
-            } else if i < lexer.newlines[0] {
-                return (0, lexer.newlines[0] - 1);
-            }
-
-            for j in 0..lexer.newlines.len() - 1 {
-                if lexer.newlines[j + 1] > i {
-                    return (lexer.newlines[j], lexer.newlines[j + 1] - 1);
-                }
-            }
-            (lexer.newlines[lexer.newlines.len() - 1], lexer.s.len())
-        }
-
-        let (st, _) = surrounding_line_off(self, span.start());
-        let (_, en) = surrounding_line_off(self, span.end());
+        let (st, _) = match self.newlines.binary_search(&span.start()) {
+            Ok(j) => (self.newlines[j], j + 1),
+            Err(0) => (0, 0),
+            Err(j) if j == self.newlines.len() + 1 => (self.newlines[j - 1], j),
+            Err(j) => (self.newlines[j - 1], j),
+        };
+        let en = match self.newlines.binary_search(&span.end()) {
+            Ok(j) if j == self.newlines.len() + 1 => self.s.len(),
```

Comment:
How can `j` ever be equal to `newlines.len() + 1`, if `binary_search` returns the index of the element when found? I assumed `j` is guaranteed to be `< newlines.len()`.